packages/react-server/src/ReactFizzServer.js

```diff
@@ -5358,14 +5394,40 @@ function flushSegment(
     return flushSubtree(request, destination, segment, hoistableState);
   }
 
+  const row = boundary.row;
+  if (
+    boundary.status === PENDING &&
+    row !== null &&
+    row.together &&
+    row.boundaries !== null &&
+    row.pendingTasks === row.boundaries.length
+  ) {
+    // If we have a "together" row and all the pendingTasks are really the boundaries themselves,
+    // and we won't outline any of them then we can unblock this row early so that we can inline
+    // all the boundaries at once.
+    let allComplete = true;
+    for (let i = 0; i < row.boundaries.length; i++) {
+      const rowBoundary = row.boundaries[i];
+      if (
+        rowBoundary.pendingTasks !== 1 ||
+        isEligibleForOutlining(request, rowBoundary)
+      ) {
+        allComplete = false;
+        break;
+      }
+    }
+    if (allComplete) {
+      unblockSuspenseListRow(request, row);
+    }
+  }
```

Comment:
This optimization wouldn't be safe if this had some hoistable state because it would not have been hoisted as part of `hoistPreambleState` since these boundaries weren't done by the time we started flushing the shell.

Not sure how to solve that.

Reply:
Probably have to find a way to move this optimization earlier.

Reply:
Moved to render phase. https://github.com/facebook/react/pull/33311/commits/1c50823b66230de9c3e52b3460495ceca17221d4